src/compose/images.coffee

```diff
@@ -0,0 +1,195 @@
+		.then (img) =>
+			if img?
+				@db.models('image').where(name: image.name).select()
+				.then (imagesFromDB) =>
+					if imagesFromDB.length == 1 and _.isEqual(@format(imagesFromDB[0]), @format(image))
+						@docker.getImage(image.name).remove(force: true)
+		.then ->
+			return removed
+
+	remove: (image) =>
+		@reportChange(image.imageId, _.merge(_.clone(image), { status: 'Deleting' }))
+		@logger.logSystemEvent(logTypes.deleteImage, { image })
+		@_removeImageIfNotNeeded(image)
+		.tap =>
+			@db.models('image').del().where(image)
+		.then (removed) =>
+			if removed
+				@logger.logSystemEvent(logTypes.deleteImageSuccess, { image })
+			else
+				@logger.logSystemEvent(logTypes.imageAlreadyDeleted, { image })
+		.catch (err) =>
+			@logger.logSystemEvent(logTypes.deleteImageError, { image, error: err })
+			throw err
+		.finally =>
+			@reportChange(image.imageId)
+
+	getNormalisedTags: (image) ->
+		Promise.map(image.RepoTags ? [], (tag) => @normalise(tag))
+
+	_withImagesFromDockerAndDB: (callback) =>
+		Promise.join(
+			@docker.listImages()
+			.map (image) =>
+				image.NormalisedRepoTags = @getNormalisedTags(image)
+				Promise.props(image)
+			@db.models('image').select()
+			callback
+		)
+
+	_isAvailableInDocker: (image, dockerImages) ->
+		_.some dockerImages, (dockerImage) ->
+			_.includes(dockerImage.NormalisedRepoTags, image.name)
+
+	# Gets all images that are supervised, in an object containing name, appId, serviceId, serviceName, imageId, dependent.
+	getAvailable: =>
+		@_withImagesFromDockerAndDB (dockerImages, supervisedImages) =>
+			return _.filter(supervisedImages, (image) => @_isAvailableInDocker(image, dockerImages))
+
+	cleanupDatabase: =>
+		@_withImagesFromDockerAndDB (dockerImages, supervisedImages) =>
+			return _.filter(supervisedImages, (image) => !@_isAvailableInDocker(image, dockerImages))
+		.then (imagesToRemove) =>
+			ids = _.map(imagesToRemove, 'id')
+			@db.models('image').del().whereIn('id', ids)
+
+	getStatus: =>
+		@getAvailable()
+		.map (image) ->
+			image.status = 'Downloaded'
+			image.download_progress = null
+			return image
+		.then (images) =>
+			status = _.clone(@volatileState)
+			_.forEach images, (image) ->
+				status[image.imageId] ?= image
+			return _.values(status)
+
+	_getDanglingAndOldSupervisorsForCleanup: =>
+		images = []
+		@docker.getRegistryAndName(constants.supervisorImage)
+		.then (supervisorImageInfo) =>
+			@docker.listImages()
+			.map (image) =>
+				Promise.map image.RepoTags ? [], (repoTag) =>
+					@docker.getRegistryAndName(repoTag)
+					.then ({ imageName, tagName }) ->
+						if imageName == supervisorImageInfo.imageName and tagName != supervisorImageInfo.tagName
+							images.push(repoTag)
+		.then =>
+			@docker.listImages(filters: { dangling: [ 'true' ] })
+			.map (image) ->
+				images.push(image.Id)
+		.then =>
+			return _.filter images, (image) =>
+				!@imageCleanupFailures[image]? or Date.now() - @imageCleanupFailures[image] > 3600 * 1000
```

Comment:
Should this timeout value be in constants?

Reply:
Moved